clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -5959,16 +5960,78 @@ static void handleIntelFPGASimpleDualPortAttr(Sema &S, Decl *D,
-  checkForDuplicateAttribute<IntelFPGAMaxReplicatesAttr>(S, D, A);
+void Sema::AddIntelFPGAMaxReplicatesAttr(Decl *D, const AttributeCommonInfo &CI,
+                                         Expr *E) {
+  if (!E->isValueDependent()) {
+    // Validate that we have an integer constant expression and then store the
+    // converted constant expression into the semantic attribute so that we
+    // don't have to evaluate it again later.
+    llvm::APSInt ArgVal;
+    ExprResult Res = VerifyIntegerConstantExpression(E, &ArgVal);
+    if (Res.isInvalid())
+      return;
+    E = Res.get();
+    // This attribute requires a strictly positive value.
+    if (ArgVal <= 0) {
+      Diag(E->getExprLoc(), diag::err_attribute_requires_positive_integer)
+          << CI << /*positive*/ 0;
+      return;
+    }
+    // Check to see if there's a duplicate attribute with different values
+    // already applied to the declaration.
+    if (const auto *DeclAttr = D->getAttr<IntelFPGAMaxReplicatesAttr>()) {
+      // If the other attribute argument is instantiation dependent, we won't
+      // have converted it to a constant expression yet and thus we test
+      // whether this is a null pointer.
+      const auto *DeclExpr = dyn_cast<ConstantExpr>(DeclAttr->getValue());
+      if (DeclExpr && ArgVal != DeclExpr->getResultAsAPSInt()) {
+        Diag(CI.getLoc(), diag::warn_duplicate_attribute) << CI;
+        Diag(DeclAttr->getLocation(), diag::note_previous_attribute);
+        return;
+      }
+    }
+    // [[intel::fpga_register]] and [[intel::max_replicates()]]
+    // attributes are incompatible.
+    if (checkAttrMutualExclusion<IntelFPGARegisterAttr>(*this, D, CI))
+      return;
+  }
 
-  if (checkAttrMutualExclusion<IntelFPGARegisterAttr>(S, D, A))
-    return;
+  D->addAttr(::new (Context) IntelFPGAMaxReplicatesAttr(Context, CI, E));
+}
+
+IntelFPGAMaxReplicatesAttr *
+Sema::MergeIntelFPGAMaxReplicatesAttr(Decl *D,
+                                      const IntelFPGAMaxReplicatesAttr &A) {
+  // Check to see if there's a duplicate attribute with different values
+  // already applied to the declaration.
+  if (const auto *DeclAttr = D->getAttr<IntelFPGAMaxReplicatesAttr>()) {
+    const auto *DeclExpr = dyn_cast<ConstantExpr>(DeclAttr->getValue());
+    const auto *MergeExpr = dyn_cast<ConstantExpr>(A.getValue());
+    if (DeclExpr && MergeExpr &&
+        DeclExpr->getResultAsAPSInt() != MergeExpr->getResultAsAPSInt()) {
+      Diag(DeclAttr->getLoc(), diag::warn_duplicate_attribute) << &A;
+      Diag(A.getLoc(), diag::note_previous_attribute);
+      return nullptr;
+    }
+  }
+  // [[intel::max_replicates()]] and [[intel::fpga_register]]
+  // attributes are incompatible.
+  if (const auto *DeclAttr = D->getAttr<IntelFPGARegisterAttr>()) {
+    Diag(DeclAttr->getLoc(), diag::err_attributes_are_not_compatible)
+        << &A << DeclAttr;
+    Diag(A.getLoc(), diag::note_conflicting_attribute);
+    return nullptr;
+  }
```

Comment:
I did not add checkAttrMutualExclusion<> here since this will generate incompatible attribute on first line instead of second line

// Merging of incompatible args
[[intel::max_replicates(12)]] extern const int var_max_replicates_2;
[[intel::fpga_register]] const int var_max_replicates_2 = 0;

Reply:
Hmm, the other `merge*Attr` use `checkAttrMutualExclusion()`, so I don't think the order is incorrect.

`D` holds the attributes we've already applied, so if `D` has an `IntelFPGARegisterAttr`, then the conflict we want to report is not with that attribute, but with the one we're about to add (`IntelFPGAMaxReplicatesAttr`). From that perspective, I think your code is diagnosing the opposite of what I'd expect -- it is diagnosing the `IntelFPGARegisterAttr` attribute location as being incompatible, but that's not the incompatible attribute -- the attempt to merge in `IntelFPGAMaxReplicatesAttr` is.

If you still think the behavior is wrong when using `checkAttrMutualExclusion<>`, can you post the exact diagnostic output you're getting from it?

Reply:
@AaronBallman, thanks for the feedback.

>>I think your code is diagnosing the opposite of what I'd expect -- it is diagnosing the IntelFPGARegisterAttr attribute location as being incompatible,

Yes, this is what happening here. 

I confused myself with CheckMutualExclusionSYCLLoopAttribute() that the incompatible diagnostic will be on second line where we have IntelFPGARegisterAttr attribute on declaration, Thanks for the explanation. I was wrong here. I will fix this and submit a new patch.

Reply:
This is the output from checkAttrMutualExclusion<> with the current patch:

// Merging of incompatible attributes
//expected-error@+2{{'max_replicates' and 'fpga_register' attributes are not compatible}}
//expected-note@+2{{conflicting attribute is here}}
[[intel::max_replicates(12)]] extern const int var_max_replicates_2;
[[intel::fpga_register]] const int var_max_replicates_2 =0;

Reply:
That order looks wrong to me. I'd expect:
```
[[intel::max_replicates(12)]] extern const int var_max_replicates_2; // expected-note {{conflicting attribute is here}}
[[intel::fpga_register]] const int var_max_replicates_2 =0; // expected-error {{'max_replicates' and 'fpga_register' attributes are not compatible}}
```
One question I have is: are you sure this behavior isn't coming from `checkIntelFPGARegisterAttrCompatibility()` rather than this change?

Reply:
>>One question I have is: are you sure this behavior isn't coming from checkIntelFPGARegisterAttrCompatibility() rather than this change?

Good question. I will take a look at this. I have not checked this function.

Reply:
Inside checkIntelFPGARegisterAttrCompatibility()
i  commented out the line
/if (checkAttrMutualExclusion<IntelFPGAMaxReplicatesAttr>(S, D, Attr))
    //InCompat = true;

I do not think this behavior is coming from this function.

I think Problem is here:

template <typename AttrTy>
static bool checkAttrMutualExclusion(Sema &S, Decl *D, const Attr &AL) {
  if (const auto *A = D->getAttr<AttrTy>()) {
    S.Diag(AL.getLocation(), diag::err_attributes_are_not_compatible) << &AL
                                                                      << A;
    S.Diag(A->getLocation(), diag::note_conflicting_attribute);
    return true;
  }
  return false;
}

This is checking wrong order. @AaronBallman, anythought?

Reply:
> That order looks wrong to me. I'd expect:
> 
> ```
> [[intel::max_replicates(12)]] extern const int var_max_replicates_2; // expected-note {{conflicting attribute is here}}
> [[intel::fpga_register]] const int var_max_replicates_2 =0; // expected-error {{'max_replicates' and 'fpga_register' attributes are not compatible}}
> ```
> 
This is what i had before checkAttrMutualExclusion<> 

  ``` // [[intel::max_replicates()]] and [[intel::fpga_register]]
  // attributes are incompatible.
  if (const auto *DeclAttr = D->getAttr<IntelFPGARegisterAttr>()) {
    Diag(DeclAttr->getLoc(), diag::err_attributes_are_not_compatible)
        << &A << DeclAttr;
    Diag(A.getLoc(), diag::note_conflicting_attribute);
    return nullptr;
  }```


```
template <typename AttrTy>
static bool checkAttrMutualExclusion(Sema &S, Decl *D, const Attr &AL) {
  if (const auto *A = D->getAttr<AttrTy>()) {
    S.Diag(AL.getLocation(), diag::err_attributes_are_not_compatible) << &AL
                                                                      << A;
    S.Diag(A->getLocation(), diag::note_conflicting_attribute);
    return true;
  }
  return false;
}```

Reply:
Thanks for checking! I don't want us to get too hung up on this, but something fishy is going on that I'm not quite understanding. Your code looks like it should issue the diagnostics in the wrong order -- you check to see if the declaration has an existing `IntelFPGARegisterAttr` attribute and if it does, it issues the diagnostic on that previous attribute (not the one being added). It then notes the one being added as being the conflicting attribute. This is backwards -- the initial attribute is fine (there were no attributes on the declaration for it to conflict with), it's the second attribute location that should be diagnosed as the error (rather than the note).

Are all of the `merge*Attr()` functions which call `checkAttrMutualExclusion` getting the order wrong? (I suppose that's a possibility I hadn't considered.) Or is it perhaps because we don't have any merge function for `IntelFPGARegisterAttr` when we should have one

Reply:
Thanks to some excellent detective work from @smanna12, we figured out that this is a community issue. All of the merge functions currently calling `checkAttrMutualExclusion()` have the same backwards diagnostic behavior. Yay for getting to the bottom of the mystery!

I think you should use `checkAttrMutualExclusion()` and add a FIXME comment to the test about this behavior not being ideal, but it being an issue that should be solved in community. I don't think we need to rush to get a fix into community because the diagnostic behavior here is sufficient for users to understand what's wrong with their code, even if the diagnostic location isn't ideal (redeclaration merging is a bit of an edge case for attribute handling).

Reply:
Thanks @AaronBallman for the comments and the detailed explanation.

Reply:
I have added FIXME comments in the test.